service/controller/resource/instance/create_drain_old_worker_nodes.go

```diff
@@ -53,7 +53,13 @@ func (r *Resource) drainOldWorkerNodesTransition(ctx context.Context, obj interf
 
 	var nodesPendingDraining int
 	for _, i := range allWorkerInstances {
-		if *i.LatestModelApplied {
+		old, err := r.isWorkerInstanceFromPreviousRelease(ctx, cr, i)
```

Comment:
I'd propose to list all nodes in the beginning and then either sort them to old and new or alternatively create a function that finds the given one from already fetched nodes. Now in here this lists all nodes from k8s api for every instance that exists in Azure VMSS. Even for small cluster it's already quite a lot of data to transfer from etcd via apiserver to client and in bigger clusters it gets exponentially worse.

Reply:
What about having worker node <-> VMSS instance mapping saved (in status?) so we can look it up when we need it?
(maybe too much/unnecessary for this fix)

Reply:
The node list in status wouldn't be guaranteed to be up-to-date-enough, I think. Consider that upgrade is started, `instance` resource proceeds to some extend, then Azure works on the background while CP gets disconnected from it. I'm not sure if that would work correctly in all cases.